overlord/fdestate/dbx.go

```diff
@@ -0,0 +1,568 @@
+					err))
+		} else {
+			reason := "aborted explicitly or due to timeout waiting for subsequent request from the caller"
+			if op.Status == AbortingStatus && op.Err != "" {
+				// aborting with explicit reason
+				reason = op.Err
+			}
+			op.SetFailed(reason)
+		}
+
+		if err := updateExternalOperation(st, op); err != nil {
+			return err
+		}
+
+		t.Logf("external action state updated to %v: %v", op.Status, op.Err)
+		return nil
+	}
+
+	return fmt.Errorf("internal error: unexpected state of external action in undo handler: %v", op.Status)
+}
+
+func (m *FDEManager) doEFISecurebootDBUpdate(t *state.Task, tomb *tomb.Tomb) error {
+	// the handler is running, which means that we are no longer blocked waiting
+	// for the op to complete
+
+	st := t.State()
+
+	st.Lock()
+	defer st.Unlock()
+
+	op, err := findFirstExternalOperationByChangeID(st, t.Change().ID())
+	if err != nil || op == nil {
+		return fmt.Errorf("internal error: cannot execute efi-dbx-update handler: %v", err)
+	}
+
+	switch op.Status {
+	case CompletingStatus:
+		// handled below
+	case AbortingStatus:
+		// explicit error when operation got aborted
+		reason := "aborted by external request"
+		if op.Err != "" {
+			// aborting with explicit reason
+			reason = op.Err
+		}
+		return errors.New(reason)
+	default:
+		return fmt.Errorf("cannot perform post update reseal, operation in status %v", op.Status)
+	}
+
+	var updateData dbxUpdateContext
+	if err := json.Unmarshal(op.Context, &updateData); err != nil {
+		return fmt.Errorf("cannot unmarshal DBX update context data: %v", err)
+	}
+
+	mgr := fdeMgr(st)
+	err = postUpdateReseal(mgr, st.Unlocker(), updateData.Method)
+	if err != nil {
+		t.Errorf("cannot complete post update reseal: %v", err)
+	}
+
+	if err != nil {
+		op.SetFailed(
+			fmt.Sprintf("cannot complete post update reseal: %v, while completing due to external request", err))
+	} else {
+		op.SetStatus(DoneStatus)
+	}
+
+	if updateErr := updateExternalOperation(st, op); updateErr != nil {
+		t.Logf("cannot update operation status: %v", updateErr)
+		return updateErr
+	}
+
+	if err == nil {
+		// update task status before unlocking
+		t.SetStatus(state.DoneStatus)
+	}
+
+	return err
+}
+
+func (m *FDEManager) doEFISecurebootDBUpdatePrepareCleanup(t *state.Task, tomb *tomb.Tomb) error {
+	st := t.State()
+
+	st.Lock()
+	defer st.Unlock()
+
+	chgID := t.Change().ID()
+	return withFdeState(st, func(fde *FdeState) (modified bool, err error) {
+		for idx, op := range fde.PendingExternalOperations {
+			logger.Debugf("cleaning up external operation for change %v", op.ChangeID)
+			if op.ChangeID == chgID {
+				fde.PendingExternalOperations = append(fde.PendingExternalOperations[:idx],
+					fde.PendingExternalOperations[idx+1:]...)
+
+				val := st.Cached(dbxUpdatePrepareSyncKey{})
+				if val != nil {
+					syncChs := val.(map[string]chan struct{})
+					delete(syncChs, chgID)
+				}
```

Comment:
should this be an helper at the end of the file?

Reply:
moved